typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,717 @@
+
+***
+</details></p>
+
+#### Let-polymorphism
+
+HM makes one concession for the sake of type inference.
+It only allows so-called let-polymorphism instead of unconstrained parametric polymorphism.
+
+To explain what this means with an example: one would like to have polymorphic type inference for expressions like
+
+$$(\lambda id.\; \ldots (id\;true) \ldots (id\;123) \ldots)\;(\lambda x.\; x),$$
+
+but type inference in polymorphic lambda calculus is in general undecidable (Wells, 1999).
+
+To work around that, HM stipulates that polymorphic types can only be inferred for let-bindings, e.g. bindings of the form
+
+$$
+\begin{array}{rl}
+\mathbf{let} & id = \lambda x.\; x\\
+\mathbf{in} & \ldots (id\;true) \ldots (id\;123) \ldots
+\end{array}
+$$
+
+Obviously, HM has to introduce the let-binding construction into lambda calculus syntax and define its semantics in terms of base lambda calculus.
+Hopefully, the translation is more or less obvious -- let-bindings are semantically equivalent to immediately-applied lambda abstractions.
+Note, however, that this doesn't yet allow for recursive definitions.
+
+#### HM syntax
+
+I will now formally introduce the syntax of expressions, types, and a function $free$ acting on type expressions.
+I'll use an informal variation of BNF (Backus-Naur form) to define syntax.
+
+The expression syntax is exactly the syntax of lambda calculus plus the let-in construct:
+
+$$
+\begin{array}{rrll}
+t & ::= & x & \text{variable}
+\\ & | & t_1\;t_2 & \text{application}
+\\ & | & \lambda x.\; t & \text{abstraction}
+\\ & | & \mathbf{let}\; x = t_1\;\mathbf{in}\;t_2 & \text{let-in}
+\end{array}
+$$
+
+Note that $t$, $t_1$, $t_2$ are recursive references.
+
+Types are introduced in two parts, the first part being the _monotypes_ $\tau$ and the second -- _polytypes_ $\sigma$ (also called _type schemes_, hence the $\sigma$):
+
+$$
+\begin{array}{rrll}
+\tau & ::= & \alpha & \text{type variable}
+\\ & | & C\;\tau_1\;\ldots\;\tau_n & \text{type constructor application}
+\\\sigma & ::= & \tau & \text{monotype}
+\\ & | & \forall \alpha .\; \sigma & \text{quantified type}
+\end{array}
+$$
+
+$C$ represents type constructors, defined as primitives, and their set is arbitrary in HM, but it must include at least the abstraction type constructor $\to$.
+
+Parentheses can be used in value and type expressions to change evaluation order, but we omit these in the syntax definitions for brevity.
+
+Note that type variables are admitted as monotypes, which makes those distinct from _monomorphic types_ that can't have type variables.
+
+Polytypes also include monotypes, in addition to quantified types.
+
+The quantifier $\forall$ _binds_ the type variables, so in a type expression $\forall \alpha_1\;\ldots\;\alpha_n.\; \tau,$ all of $\alpha_i$ are _bound_.
+Type variables that are not bound are called _free_.
+
+Additionally, type variables may be bound by the typing context.
```

Comment:
This sounds a bit odd to me. The typing context exists at a meta-level, whereas I'd expect variable bindings to occur directly in the syntax. However, I'm not very familiar with HM. I suppose I just don't know enough about HM to review this section of the article – saying this just for the record.

Reply:
System F is introduced later in the article and it binds type variables the way I expected. Apparently my expectations were shaped by the way GHC does things internally, and it's based on System F, so that makes sense.

Reply:
You're right, it does sound weird. Poor choice of words on my part. It would be more accurate to say that type variables are considered bound if they occur in the context.